src/pkgs/files/login.nix

```diff
@@ -1,29 +1,85 @@
 # Licensed under GNU Lesser General Public License v3 or later, see COPYING.
 # Copyright (c) 2019 Alexander Sosedkin and other contributors, see AUTHORS.
 
-{ instDir, writeScript }:
+{ instDir, writeScriptBin }:
 
-writeScript "login" ''
+writeScriptBin "login" ''
   #!/system/bin/sh
   set -e
 
-  export USER=$(/system/bin/whoami)
+  ORIGINAL_PATH="$PATH"
+  ANDROID_LIKELY_PATH="/sbin:/system/sbin:/system/bin:/system/xbin"
+  ANDROID_LIKELY_PATH="$ANDROID_LIKELY_PATH:/vendor/bin:/vendor/xbin"
+  PATH="$ANDROID_LIKELY_PATH:$OLD_PATH"
+
+  resolve_link() {
```

Comment:
Oh boy, this is really complicated.. Can you explain in words what exactly was the problem and we do you need so many cases?

AFAICT the symlinks of login (failing in the android app) and proot-static, sh and login-inner are the problem right? So the symlink blob in the `~/.nix-profile` is the real source of problems, I see.. I need a moment to work this through :D

Reply:
Ha-ha, it totally is.

Short explanation: the real files are hidden under layers and layers of symlinks. This would not be a problem at all if only those were working symlinks. They are not.

Complication 1: symlinks can point to, for example, `/nix/store/123-whatever`. This doesn't make sense for the outside Android system. You see such a symlink (not pointing to `/data/data/com.termux.nix/files`), you prepend `/data/data/com.termux.nix/files/usr` to it. Now it works.

Complication 2: symlinks can eventually point to `/long/path/that/has/a/broken_symlink/and/something/after/it`. When you have this thing, you have to iterate over it and see which one of the parents is a broken symlink, fix it, resolve it, add the remainder and continue resolving. It's bonkers.

Here `login` finds `proot-static`. There's also a Java-land implementation of this that finds `login`: https://github.com/t184256/nix-on-droid-app/commit/f4d6e55cda63735a504669fbdf0c1fdd50cbe657

You can marvel at my notes on how `login` is found:

```
    // /d/d/c.t.n/f/usr/bin/login -> /d/d/c.t.n/f/usr/nix/v/n/p/p/USERNAME/profile/login
    // /d/d/c.t.n/f/usr/nix/v/n/p/p/USERNAME/profile/login is broken
    // /d/d/c.t.n/f/usr/nix/v/n/p/p/USERNAME/profile -> profile-1-link
    // /d/d/c.t.n/f/usr/nix/v/n/p/p/USERNAME/profile-1-link -> /nix/store/...-user-environment
    // /nix/store/...-user-environment/bin
    // /d/d/c.t.n/f/usr/nix/store/...-user-environment/bin -> /nix/store/...-basic-environment/bin
    // /d/d/c.t.n/f/usr/nix/store/...-basic-environment/bin/login -> /nix/store/...-login/bin/login
```

or uncomment the debugging output in the shell script and see a whole screen of detective work to find `proot-static`.

Also, shipping `login` that is a broken symlink broke my tablet setup, actually, as I have an external process (Easer) that gains root, enters the Nix environment and executes an evdev/uinput keyboard remapper when I connect my Bluetooth keyboard. So, maybe one day I'll move this symlink resolver into a separate file to reuse it in other 'Android chroots into Nix' scenarios.

But hey, now we have an upgrade path that doesn't require wiping. Worth it.